File: News_Service_Jooq/src/main/java/com/dunice/projectNews/services/impl/TagServiceImpl.java
package com.dunice.projectNews.services.impl;

import com.dunice.projectNews.dto.request.NewsRequest;
import com.dunice.projectNews.models.Tag;
import com.dunice.projectNews.repository.TagRepository;
import com.dunice.projectNews.services.TagService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Set;

@Service
@RequiredArgsConstructor
public class TagServiceImpl implements TagService {

    private final TagRepository tagRepository;

    @Transactional
    public Set<Long> parseTag(NewsRequest newsRequest) {

        return tagRepository.saveNewsWithTags(newsRequest);
    }

}

File: News_Service_Kotlin/src/main/kotlin/com/dunice/advancedServiceKotlin/services/impl/TagServiceImpl.kt
package com.dunice.advancedServiceKotlin.services.impl

import com.dunice.advancedServiceKotlin.dto.request.NewsRequest
import com.dunice.advancedServiceKotlin.exception.CustomException
import com.dunice.advancedServiceKotlin.models.Tag
import com.dunice.advancedServiceKotlin.repository.TagRepository
import com.dunice.advancedServiceKotlin.services.TagService
import com.dunice.advancedServiceKotlin.util.ServerErrorCodes
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.stream.Collectors

@Service
class TagServiceImpl @Autowired constructor(private val tagRepository: TagRepository) : TagService {
    @Transactional
    override fun parseTag(newsRequest: NewsRequest): Set<Tag> {
        val tagsForNonExistent = mutableSetOf<Tag>()
        val tags = newsRequest.tags ?: throw CustomException(ServerErrorCodes.TASK_NOT_FOUND)

        val tagsSet = tags.map { s: String? ->
            tagRepository.findByTitle(s)?.orElseGet {
                val newTag = Tag(title = s ?: "")
                tagsForNonExistent.add(newTag)
                newTag
            }?: throw CustomException(ServerErrorCodes.TASK_NOT_FOUND)
        }.toSet()

        if (tagsForNonExistent.isNotEmpty()) {
            tagRepository.saveAll(tagsForNonExistent)
        }

        return tagsSet
    }

}
